editor/py/serve.py

```diff
@@ -375,11 +386,52 @@ def resolve_tokens(proto):
   return matched
 
 
-def init_db():
-  """Ensures the db/ directory exists and contains at least one Dataset."""
-  if not os.path.isdir('db'):
-    os.mkdir('db')
-  if os.listdir('db'):
-    return
-  dataset = dataset_pb2.Dataset()
-  put_dataset('dataset', dataset)
+@app.before_request
+def init_user():
+  """Set the user cookie and initialize the db/{user} directory."""
+  user = flask.request.cookies.get('ord-editor-user')
+  if user is None:
+    user = flask.request.args.get('user') or next_user()
+    return redirect_to_user(user)
+  flask.g.user = user
+  if not os.path.isdir(f'db/{user}'):
+    os.mkdir(f'db/{user}')
+  if not os.listdir(f'db/{user}'):
+    dataset = dataset_pb2.Dataset()
+    put_dataset('dataset', dataset)
+
+
+def ensure_user():
+  """On page requests only, a user param in the URL overrides the cookie."""
+  url_user = flask.request.args.get('user')
+  cookie_user = flask.request.cookies.get('ord-editor-user')
+  # If there is no user in the URL, set it from the cookie.
+  if url_user is None:
+    return redirect_to_user(cookie_user)
+  # If the cookie and the URL disagree, the URL wins.
+  if url_user != cookie_user:
+    return redirect_to_user(url_user)
+  # If the URL and the cookie are consistent, then do nothing.
+
+
+def redirect_to_user(user):
+  """Set the user cookie and return a redirect to the updated URL."""
+  url = url_for_user(user)
+  response = flask.redirect(url)
+  # Expires in a year.
+  response.set_cookie('ord-editor-user', user, max_age=31536000)
+  return response
+
+
+def url_for_user(user):
+  parts = list(urllib.parse.urlparse(flask.request.url))
+  parts[4] = urllib.parse.urlencode({'user': user})
+  return urllib.parse.urlunparse(parts)
+
+
+def next_user():
+  """Return a user identifier that is not present in the db/ directory."""
+  user = ''.join(random.choice(string.ascii_letters) for i in range(4))
```

Comment:
Can we use a full `uuid.uuid4().hex` here? I want to reduce the risk of guessing a valid user ID and messing with their data.

Reply:
Done.